doc/man7/provider-base.pod

```diff
@@ -0,0 +1,375 @@
+=pod
+
+=head1 NAME
+
+provider-base
+- The basic OpenSSL library E<lt>-E<gt> provider functions
+
+=head1 SYNOPSIS
+
+ #include <openssl/core_numbers.h>
+
+ /*
+  * None of these are actual functions, but are displayed like this for
+  * the function signatures for functions that are offered as function
+  * pointers in OSSL_DISPATCH arrays.
+  */
+
+ /* Functions offered by libcrypto to the providers */
+ const OSSL_ITEM *core_get_param_types(const OSSL_PROVIDER *prov);
+ int core_get_params(const OSSL_PROVIDER *prov, OSSL_PARAM params[]);
+ int core_thread_start(const OSSL_PROVIDER *prov,
+                       OSSL_thread_stop_handler_fn handfn);
+ void core_put_error(const OSSL_PROVIDER *prov,
+                     uint32_t reason, const char *file, int line);
+ void core_add_error_vdata(const OSSL_PROVIDER *prov, int num, va_list args);
+ OPENSSL_CTX *core_get_library_context(const OSSL_PROVIDER *prov);
+
+ /* Some OpenSSL functionality is directly offered to providers via dispatch */
+ void *CRYPTO_malloc(size_t num, const char *file, int line);
+ void *CRYPTO_zalloc(size_t num, const char *file, int line);
+ void *CRYPTO_memdup(const void *str, size_t siz, const char *file, int line);
+ char *CRYPTO_strdup(const char *str, const char *file, int line);
+ char *CRYPTO_strndup(const char *str, size_t s, const char *file, int line);
+ void CRYPTO_free(void *ptr, const char *file, int line);
+ void CRYPTO_clear_free(void *ptr, size_t num, const char *file, int line);
+ void *CRYPTO_realloc(void *addr, size_t num, const char *file, int line);
+ void *CRYPTO_clear_realloc(void *addr, size_t old_num, size_t num, const char *file, int line);
+ void *CRYPTO_secure_malloc(size_t num, const char *file, int line);
+ void *CRYPTO_secure_zalloc(size_t num, const char *file, int line);
+ void CRYPTO_secure_free(void *ptr, const char *file, int line);
+ void CRYPTO_secure_clear_free(void *ptr, size_t num, const char *file, int line);
+ int CRYPTO_secure_allocated(const void *ptr);
+ void OPENSSL_cleanse(void *ptr, size_t len);
+ unsigned char *OPENSSL_hexstr2buf(const char *str, long *len);
```

Comment:
Side note: This seems like a very strange function for us to be providing like this

Reply:
It does memory allocation, that's the reason it's bunched in like this. I kinda [shrug], ya know?

Reply:
Well...yeah....it calls OPENSSL_malloc...but so do lots of other functions in the library - so why this one? Anyway this was a side note.

Reply:
That à question to ask @richsalz

Reply:
Poof, summon me and I appear :)  The answer is in the message for commit b60cba3c5d25dcd5cea2c97193a30b730c8a0f28

Reply:
That doesn't seem to answer the question. Just because EVP control functions need hexstring conversion doesn't seem like a good reason to make this function available to all providers.

Reply:
That's a discussion for another PR.  This PR documents what's already present.

Reply:
Yes...this was always a side note.

Reply:
I don't want to hijack this PR, if there's another place to discuss this let me know.  EVP uses hexstring, EVP is in the boundary, therefore hexstring needs to be in the boundary as well. I remain confused by Matt's concern.

Reply:
There's no real need to export OPENSSL_hexstr2buf(), the providers that use EVP could use the source of that function directly, as long as they make sure to have OPENSSL_malloc() and friends available within their boundaries *somehow*

Reply:
Exactly. OPENSSL_hexstr2buf is just a utility function. We don't want to bloat the libcrypto<->provider interface with a whole load of utility functions. IMO, we only want to provide the bare essential "OS" level functions that providers couldn't easily provide themselves. If providers really want to take advantage of a broader set of utility functions made available by libcrypto - then they can just link against libcrypto.

Reply:
EVP requires hexstring conversion.

I did the simplest thing that fixed the build after EVP was brought inside the boundary.  Feel free to improve it, of course.

Reply:
#9432